engine/runtime/src/main/java/org/enso/interpreter/runtime/ResourceManager.java

```diff
@@ -158,7 +162,7 @@ public void shutdown() {
       Item it = items.remove(key);
       if (it != null) {
         // Finalize unconditionally – all other threads are dead by now.
-        it.doFinalize(context);
+        it.scheduleFinalizeAtSafepoint(context);
```

Comment:
Are we guaranteed a Truffle safepoint during context finalization? If not, we should probably have a separate path for finalization to make sure we clean everything up?

Reply:
Thanks for the comment. I restructured the code in 545b89c:
- on `Context.close()` I directly invoke `finalizeNow` - I can do it, as we are "in the context" and can execute the code directly
- when a resource is obtained from `ReferenceQueue`, I schedule the safepoint check - that get's me into the context as soon as some Enso code is executed

If the tests pass, I integrate.